Configurable rate limit for GitHub Actions on GHES (#16945)
github/docs: content/admin/configuration/configuring-your-enterprise/configuring-rate-limits.md

3. Type limits for Total Requests, CPU Limit, and CPU Limit for Searching, or accept the pre-filled default limits.
{% data reusables.enterprise_management_console.save-settings %}

> {% ifversion ghes > 3.4 %}

Author: whoops, yeah definitely

### About rate limits for {% data variables.product.prodname_actions %}

Your {% data variables.product.product_name %} instance assigns each {% data variables.product.prodname_actions %} workflow job to a runner. If your instance cannot immediately assign a job to an available runner, the job will wait in a queue until a runner is available. If {% data variables.product.prodname_actions %} experiences sustained high load, the queue can back up, and the performance of {% data variables.product.product_location %} may degrade.

To avoid this performance degradation, you can configure a rate limit for {% data variables.product.prodname_actions %}. This rate limit is expressed in job runs per minute. {% data variables.product.product_name %} calculates and applies the rate limit for the sum total of all job runs on the instance. If runs exceed the rate limit, additional runs will fail instead of entering the queue.

> ![Screenshot of a rate-limited workflow run](/assets/images/enterprise/actions/actions-rate-limited.png)

Reviewer: Maybe just a text representation of the error would be more useful than this screenshot (the error text is really small in the image), and it would also have the added benefit of being searchable both for SEO and our site search.

> To avoid this performance degradation, you can configure a rate limit for {% data variables.product.prodname_actions %}. This rate limit is expressed in job runs per minute. {% data variables.product.product_name %} calculates and applies the rate limit for the sum total of all job runs on the instance. If runs exceed the rate limit, additional runs will fail instead of entering the queue.
> 
> ![Screenshot of a rate-limited workflow run](/assets/images/enterprise/actions/actions-rate-limited.png)

Reviewer 2: ```suggestion
To avoid this performance degradation, you can configure a rate limit for {% data variables.product.prodname_actions %}. This rate limit is expressed in job runs per minute. {% data variables.product.product_name %} calculates and applies the rate limit for the sum total of all job runs on the instance. If runs exceed the rate limit, additional runs will fail instead of entering the queue. The following error will appear in the run's annotations.

> You've exceeded the rate limit for workflow run requests. Please wait before retrying the run.
```

By default, the rate limit for {% data variables.product.prodname_actions %} is disabled. Because {% data variables.product.product_name %} can handle temporary spikes in usage without performance degradation, this rate limit is intended to protect against sustained high load. We recommend leaving the rate limit disabled unless you are experiencing performance problems. In some cases, {% data variables.contact.github_support %} may recommend that you enable a rate limit for {% data variables.product.prodname_actions %}. 

### Enabling or disabling rate limits for {% data variables.product.prodname_actions %}

{% data reusables.enterprise_installation.ssh-into-instance %}
1. To enable and configure the rate limit, run the following two commands, replacing **RUNS-PER-MINUTE** with the value of your choice.

>    ```
>    ghe-config actions-rate-limiting.enabled true
>    ghe-config actions-rate-limiting.queue-runs-per-minute <em>RUNS-PER-MINUTE</em>
>    ```

Reviewer: This code block needs to be specified as `shell`, otherwise the `<em>` appears literally:

![image](https://user-images.githubusercontent.com/4434330/163904098-b7e08668-9c5f-4ae0-b55d-03b5183c4e64.png)

````suggestion
   ```shell
   ghe-config actions-rate-limiting.enabled true
   ghe-config actions-rate-limiting.queue-runs-per-minute <em>RUNS-PER-MINUTE</em>
   ```
````

> 1. Or, to disable the rate limit once it's been enabled, run the following command.

Author: @mattpollard I tweaked this line slightly from your suggestion

Reviewer 2: 👍🏻 Nice, works for me.

```suggestion
1. To disable the rate limit after it's been enabled, run the following command.
```

3. Type limits for each repository network or user ID.
  ![Fields for repository network and user ID limits](/assets/images/enterprise/management-console/example-git-rate-limits.png)
{% data reusables.enterprise_management_console.save-settings %}

## Enabling Actions rate limits

> You can apply a rate limit to GitHub Actions workflow runs. The Actions rate limit is expressed in runs per minute. It is for the sum total of all Actions runs across the entire GHES instance.

Reviewer 3: I don't love "workflow runs for GitHub Actions". I think people talk about "workflow runs for my CI/CD" or "workflow runs for other job to be done".

I think that something like the original "GitHub Actions workflow runs" is a little less ambiguous. "workflow runs executed by GitHub Actions" is also a possibility. 🤔 

Optional feedback, of course.

> You can apply a rate limit to GitHub Actions workflow runs. The Actions rate limit is expressed in runs per minute. It is for the sum total of all Actions runs across the entire GHES instance.

Author: I do get the point about stacked modifiers.  But I agree with @ethomson on this.  I don't think there's any chance for ambiguity by saying "GitHub Actions workflow runs" and I think it's the clearer phrasing.

> By default, the rate limit is disabled. When workflow jobs cannot immediately be assigned to an available runner, they will wait in a queue until a runner is available. However, if Actions experiences a sustained high load, the queue can back up faster than it can drain and the performance of the GHES instance may degrade.

Reviewer 3: Similarly here for "workflow job for GitHub Actions" in the first sentence.

To avoid this, an administrator can configure a rate limit. From the [administrative shell (SSH)](/enterprise/{{ currentVersion }}/admin/guides/installation/accessing-the-administrative-shell-ssh/), run the following commands:

```
actions-rate-limiting.enabled true
actions-rate-limiting.queue-runs-per-minute 200
ghe-config-apply
```

 When the rate limit is exceeded, additional workflow runs will fail immediately rather than being put in the queue:

> ![screenshot of a rate-limited run](/assets/images/enterprise/actions/actions-rate-limited.png)

Author: I took off the `$` in the first block because the other ones don't have it.  I also don't like when code snippets have that because it messes up copy-paste.  But happy to go with whatever our convention is.

> I think the `ghe-config` command is missing for the first two configuration values

I'm pretty sure I didn't use that when I tested it because I just copied the same snippet here 🤔
Now I don't know why that should work.  Having `ghe-config` shouldn't hurt, but I can double check.
